### PR TITLE
PCODAR: Add Point Cloud mapping package

### PR DESCRIPTION
**Pull request by kev-the-dev**
depends on #140

Major refactor / feature additions for PCODAR, the as of yet unmerged point cloud mapping package started by Tess (fall 2017) and continued by Daniel (spring 2017).

### CODE REVIEW

File: perception/point_cloud_object_detection_and_recognition/tools/labeling_tool/main.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+#include <QApplication>
+#include <QDesktopWidget>
+#include <QStyle>
+
+#include "label_model.hpp"
+#include "main_window.hpp"
+
+/*
+UI Features I want:
```
Comment (Grymessa) [inline perception/point_cloud_object_detection_and_recognition/tools/labeling_tool/main.cpp]: Found UI are these still features we want and do we use this?

File: perception/point_cloud_object_detection_and_recognition/tools/labeling_tool/label_model.hpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+#include <Eigen/Dense>
+
+#include <map>
+#include <memory>
+
+#include <point_cloud_object_detection_and_recognition/pcodar_types.hpp>
+
+#include <mil_msgs/PerceptionObjectArray.h>
+
+// namespace pcod
```
Comment (Grymessa) [inline perception/point_cloud_object_detection_and_recognition/tools/labeling_tool/label_model.hpp]: Dead Code?

File: perception/point_cloud_object_detection_and_recognition/tools/labeling_tool/label_model.cpp [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+  for (const auto& id_object_pair : *object_map_ptr_)
+  {
+    objects.push_back(id_object_pair.second.p_object);
+  }
+
+  const int number_pcloud_messages = view.size() - view_odom.size();
+
+  std::cout << "\nSAVING BAG" << std::endl;
+
+  // TODO(tbianchi) Make this a popup thingy
```
Comment (Grymessa) [inline perception/point_cloud_object_detection_and_recognition/tools/labeling_tool/label_model.cpp]: This a feature we need or still want?

File: perception/point_cloud_object_detection_and_recognition/tools/labeling_tool/label_model.cpp [hunk 4]
```diff
@@ -0,0 +1,399 @@
+#include "label_model.hpp"
+#include "main_window.cpp"
```
Comment (Grymessa) [inline perception/point_cloud_object_detection_and_recognition/tools/labeling_tool/label_model.cpp]: So the idea of this file is to label bags with PCODAR clusters so we know what they are in the future? This opens up some interesting possibilities for machine learning down the road. Were there any other plans for this? To use this to label clusters in real time?

**Review by Grymessa — CHANGES_REQUESTED** [review]
Just a few things in that label script, possible dead code and old TODOs. Really well made overall and I'd like to see it working on test day tomorrow! On a side note, this would really benefit from a README or more detailed documentation so its quicker to read through and understand what is going on. Just descriptions of the general flow and architecture of this package. As it stands its not too bad to figure out. Just would be a nice thing to have for future members.

**Comment by kev-the-dev** [comment]
Most of your comments concerned the pointcloud labeling utilities, which I have not modified from the original work by Tess. I've moved this code to deprecated to potentiality be revived later.

The remaining code has been extensively tested. It  does not adequately remove noise for NaviGator's needs (false positives for new objects), but that can be improved with future work. I'd like to merge PCODAR with the current set of features to fix version control dependencies from the other PRs depending on this. 

I have updated the README with detailed description, usage guide, and features TODOs.

**Review by Grymessa — APPROVED** [review]
Sounds good then, thanks for the updates on the README and such.

**Pull request merged**